benches/assign_ops.rs

```diff
@@ -14,7 +14,7 @@ fn add_benchmark(c: &mut Criterion) {
         c.bench_function(&format!("apply_mul 2^{}", log2_size), |b| {
             b.iter(|| {
                 criterion::black_box(&mut arr_a)
-                    .apply_values(|x| x.iter_mut().for_each(|x| *x *= 1.01));
+                    .apply_values_mut(|x| x.iter_mut().for_each(|x| *x *= 1.01));
```

Comment:
`apply_values_mut` will do the memcopy no matter what state.

I've added an `apply_values` which let's the caller decide if the the mutable path is worth it. Both functions clean up and check if the invariants hold.